Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/DriverFeeUpdates/DriverFee.hs

```diff
@@ -0,0 +1,119 @@
+import Storage.Queries.DriverFee
+import Storage.Queries.DriverInformation (updateSubscription)
+import qualified Storage.Queries.Person as QPerson
+import qualified Tools.Notifications as Notify
+
+sendPaymentReminderToDriver ::
+  ( MonadFlow m,
+    HedisFlow m r,
+    CoreMetrics m,
+    HasCacheConfig r,
+    EsqDBFlow m r,
+    Esq.EsqDBReplicaFlow m r
+  ) =>
+  Job 'SendPaymentReminderToDriver ->
+  m ExecutionResult
+sendPaymentReminderToDriver Job {id, jobInfo} = withLogTag ("JobId-" <> id.getId) do
+  let jobData = jobInfo.jobData
+      startTime = jobData.startTime
+      endTime = jobData.endTime
+  (driverFees, driverFeeIds, feeZipDriver, _) <- calcDriverFeeAttr ONGOING startTime endTime
+  if not $ null driverFees
+    then do
+      for_ feeZipDriver $ \(driverFee, mbDriver) -> do
+        case mbDriver of
+          Nothing -> do
+            logInfo "Driver Not found. This should not be possible."
+            throwError (InternalError "Driver Not Found") -- Unreachable
+          Just driver -> do
+            let paymentTitle = "Payment Pending"
+                paymentMessage = "Payment of Rs. " <> show driverFee.totalAmount.getMoney <> " is pending. Pay by " <> show driverFee.payBy <> " to avoid being blocked."
+            Notify.sendNotificationToDriver driver.merchantId FCM.SHOW Nothing FCM.PAYMENT_PENDING paymentTitle paymentMessage driver.id driver.deviceToken
+      Esq.runTransaction $ do
+        mapM_ (updateStatus PAYMENT_PENDING) driverFeeIds
+      ReSchedule <$> getRescheduledTime
+    else do
+      logInfo "No ongoing payment found."
+      return Complete
+
+unsubscribeDriverForPaymentOverdue ::
+  ( MonadFlow m,
+    HedisFlow m r,
+    CoreMetrics m,
+    HasCacheConfig r,
+    EsqDBFlow m r,
+    Esq.EsqDBReplicaFlow m r
+  ) =>
+  Job 'UnsubscribeDriverForPaymentOverdue ->
+  m ExecutionResult
+unsubscribeDriverForPaymentOverdue Job {id, jobInfo} = withLogTag ("JobId-" <> id.getId) do
+  let jobData = jobInfo.jobData
+      startTime = jobData.startTime
+  now <- getCurrentTime
+  (driverFees, driverFeeIds, feeZipDriver, relevantDriverIds) <- calcDriverFeeAttr PAYMENT_PENDING startTime now
+  if not $ null driverFees
+    then do
+      for_ feeZipDriver $ \(driverFee, mbDriver) -> do
+        case mbDriver of
+          Nothing -> do
+            logInfo "Driver Not found. This should not be possible."
+            throwError (InternalError "Driver Not Found") -- Unreachable
+          Just driver -> do
+            let paymentTitle = "Blocked - Payment Overdue"
+                paymentMessage = "Sorry, you are blocked from taking rides as payment of Rs. " <> show driverFee.totalAmount.getMoney <> " is outstanding. Please pay the balance to get unblocked."
+            Notify.sendNotificationToDriver driver.merchantId FCM.SHOW Nothing FCM.PAYMENT_OVERDUE paymentTitle paymentMessage driver.id driver.deviceToken
+      Esq.runTransaction $ do
+        mapM_ (updateStatus PAYMENT_OVERDUE) driverFeeIds
```

Comment:
can we run transaction for each driver <> payment status pair separately, instead of running all these in a single transaction block ?

Reply:
Won't it create inconsistency if one fails?

Reply:
here we are updating multiple `driver + driverFee`  pairs, right ? then each pair individually should have consistency, if one of the query fails for one pair the both should be rolled back, but its not required across pairs right ?